crates/runtime/src/lib.rs

```diff
@@ -304,14 +333,18 @@ impl Runtime {
     }
 
     pub async fn start_pods_watcher(&mut self) -> notify::Result<()> {
-        let mut current_app = Arc::clone(&self.app);
-
         let mut rx = self.pods_watcher.watch()?;
 
         while let Some(new_app) = rx.recv().await {
+            let current_app = self.app.read().await;
+            let current_datasets = current_app.datasets.clone();
+            let current_models = current_app.models.clone();
+
             tracing::debug!("Updated pods information: {:?}", new_app);
             tracing::debug!("Previous pods information: {:?}", current_app);
 
+            drop(current_app);
```

Comment:
Did you mean to drop the app here?

Reply:
This is intentionally as we would like to update the app later (acquire write lock). Don't get write lock immediately as load/modify functions for datasets and models could be modified in the future and would like to access app state which will result in the deadlock.

Reply:
As you have it structured now this introduces a race condition. If I add 2 new datasets in quick succession then the first will read the apps/models and drop the app. If the 2nd dataset reads the existing apps and models before the 1st sets the new app then it will try to read a non-existent app. Same goes for anything that is trying to read during the window between the old app being dropped and the new one being set.

Reply:
You're basically defeating the purpose of the Arc which keeps a reference count to ensure the app isn't dropped while something still needs it.

Reply:
If you want to ensure the models and datasets aren't changed after this code reads the models/datasets then you need to acquire the write lock instead. That won't cause deadlock as long as you release the lock once you're done updating.

Reply:
I misread this, you're dropping the lock, not the app itself.